language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs

```diff
@@ -77,45 +73,117 @@ import           Data.HashMap.Monoidal
+
+-- 'Values' for the modified CEK machine.
+data CekVal uni =
+    -- TODO: we probably want to store a @Some (ValueOf uni)@ here, but then we have trouble in
+    -- 'readKnownCek'. I'll reconsider the way we deal with annotations once again.
+    VCon (TermWithMem uni)
+  | VTyAbs ExMemory TyName KindWithMem (TermWithMem uni) (CekValEnv uni)
+  | VLamAbs ExMemory Name (TypeWithMem uni) (TermWithMem uni) (CekValEnv uni)
+  | VIWrap ExMemory (TypeWithMem uni) (TypeWithMem uni) (CekVal uni)
+  | VBuiltin
+      ExMemory
+      (CekValEnv uni)   -- Initial environment, used for evaluating every argument
+      StagedBuiltinName
+      Int               -- Number of arguments to be provided
+      [TypeWithMem uni] -- The types the builtin is to be instantiated at.
+                        -- We need these to construct a term if the machine is returning a stuck partial application.
+      [CekVal uni]      -- Arguments we've computed so far.
+    deriving (Show, Eq) -- Eq is just for tests.
+
+instance (Closed uni, GShow uni, uni `Everywhere` PrettyConst) => PrettyBy PrettyConfigPlc (CekVal uni) where
+    prettyBy cfg = prettyBy cfg . dischargeCekVal
+
+type instance UniOf (CekVal uni) = uni
+
+instance (Closed uni, uni `Everywhere` ExMemoryUsage) => FromConstant (CekVal uni) where
+    fromConstant = VCon . fromConstant
+
+instance AsConstant (CekVal uni) where
+    asConstant (VCon term) = asConstant term
+    asConstant _           = Nothing
+
+instance ToExMemory (CekVal uni) where
+    toExMemory = \case
+        VCon t -> termAnn t
+        VTyAbs ex _ _ _ _ -> ex
+        VLamAbs ex _ _ _ _ -> ex
+        VIWrap ex _ _ _ -> ex
+        VBuiltin ex _ _ _ _ _ -> ex
+
+type CekValEnv uni = UniqueMap TermUnique (CekVal uni)
+
+-- See Note [Scoping].
+-- | Instantiate all the free variables of a term by looking them up in an environment.
+dischargeCekValEnv :: CekValEnv uni -> TermWithMem uni -> TermWithMem uni
+dischargeCekValEnv valEnv =
+    -- We recursively discharge the environments of Cek values, but we will gradually end up doing
+    -- this to terms which have no free variables remaining, at which point we won't call this
+    -- substitution function any more and so we will terminate.
+    termSubstFreeNames $ \name -> do
+        val <- lookupName name valEnv
+        Just $ dischargeCekVal val
+
+unstageBuiltinName :: ann -> StagedBuiltinName -> Builtin ann
+unstageBuiltinName ann (StaticStagedBuiltinName name)  = BuiltinName ann name
+unstageBuiltinName ann (DynamicStagedBuiltinName name) = DynBuiltinName ann name
+
+mkBuiltinApp :: ExMemory -> Builtin ExMemory -> [TypeWithMem uni] -> [TermWithMem uni] -> TermWithMem uni
+mkBuiltinApp ex bn tys args = mkIterApp ex (mkIterInst ex (Builtin ex bn) tys) args
+
+-- Convert a CekVal into a term by replacing all bound variables with the terms
+-- they're bound to (which themselves have to be obtain by recursively discharging values).
+dischargeCekVal :: CekVal uni -> TermWithMem uni
```

Comment:
Perhaps it's worth keeping this one close to `dischargeCekValEnv` so it's clear they're mutually recursive.

Reply:
Yes, I'll do that.  `unstageBuiltinName` will go away when I get rid of staged names anyway.  I was also tempted to put (something like) `mkBuiltinApp` in `Apply.hs` since we have to do that in a number of places, but maybe that doesn't make sense if we're going to have interleaved term/type arguments.